flips/20220708-resource-reference-invalidation.md

```diff
@@ -0,0 +1,179 @@
+In other words, references stay alive forever.
+
+For example, consider the following resource.
+
+```cadence
+resource R {
+    pub(set) var id: Int
+
+    init() {
+        self.id = 1
+    }
+}
+```
+
+Now, create an instance from the above resource, then take a reference to it, and finally transfer it to an account.
+The transferred resource can still be accessed via the reference taken prior to the transfer.
+
+```cadence
+// Create a resource.
+let r <-create R()
+
+// And take a reference.
+let ref = &r as &R
+
+// Then transfer the resource into an account.
+account.save(<-r, to: /storage/r)
+
+// Update the reference.
+// This will also update the referring resource in the account.
+ref.id = 2
+```
+
+In the above example, resource `r` was initially created on the stack and was not owned by anyone.
+But similarly, it is also possible for someone to take a reference to a resource in their account, and transfer the
+resource to someone, and keep accessing the transferred resource through the reference.
+```cadence
+// Create a resource collection in 'foo' account
+fooAccount.save(<-[<-create R()], to: /storage/a)
+
+// Take a ephemeral reference to the resource at index '0' in the collection.
+// Resource is in the 'foo' account at the time of taking the reference.
+let collectionRef = fooAccount.borrow<&[R]>(from: /storage/a)!
+let ref = &collectionRef[0] as &R
+
+// Transfer the entire collection to the 'bar' account.
+let collection <- fooAccount.load<@[R]>(from: /storage/a)!
+barAccount.save(<- collection, to: /storage/b)
+
+// Can still access the transferred resource through the reference.
+ref.id = 2
+```
+
+## User Benefit
+This proposal would avoid a potential foot-gun of gaining/giving/retaining unintended access to resources through
+references.
+
+## Implementation Benefit
+The current implementation of Cadence resource reference tracking requires a lot of bookkeeping in order to ensure
+that the references to resources point to the correct resource, when the referred-to resource value is moved around.
+With this proposal, such bookkeeping would no longer be required and the implementation complexity could be eliminated.
+
+## Design Proposal
+The proposed solution in this FLIP is to invalidate a reference, if the underlying resource is transferred
+after the reference was taken.
+The reference is invalidated upon the first transfer, regardless of the origin and the destination.
+
+There can be different types of transfers, depending on the origin (where the reference was obtained), and the
+destination.
+
+| Origin  | Destination       | Reference validity |
+|:--------|:------------------|:-------------------|
+| Stack   | Stack             | Invalid            |
+| Stack   | Account           | Invalid            |
+| Account | Stack             | Invalid            |
+| Account | Same Account      | Invalid            |
+| Account | Different Account | Invalid            |
```

Comment:
Given that all cases are now invalid, maybe remove the third column, and maybe just make this a list